llama.cpp

```diff
@@ -14408,8 +14421,8 @@ void llama_sample_grammar(struct llama_context * ctx, llama_token_data_array * c
     candidates_grammar.reserve(candidates->size);
 
     for (size_t i = 0; i < candidates->size; ++i) {
-        const llama_token id    = candidates->data[i].id;
-        const std::string piece = llama_token_to_piece(ctx, id, false);
+        const llama_token id      = candidates->data[i].id;
+        const std::string & piece = ctx->model.vocab.cache_token_to_piece.at(id);
```

Comment:
This is not checking for the presence of a cache with `ctx->model.vocab.has_cache` -- should it?

Reply:
Comment at https://github.com/ggerganov/llama.cpp/pull/7587#discussion_r1619293050 resolved this question.